### PR TITLE
common: resolve config proxy deadlock using refcounted pointers

### PR DESCRIPTION
**Pull request by batrick**
7e8c683 introduced some gymnastics with a "CallGate" to maintain a count for each observer we may be "calling into" with a config change (namely: handle_conf_change). This was to prevent remove_observer coming in and deleting the observer in the middle of the call. More importantly, it was to avoid holding the lock while traversing the observers so that the config_proxy lock can be dropped while calling handle_conf_change. This is important as e.g. the MDS may attempt to acquire the config_proxy lock in its MDSRank::handle_conf_change method (what prompted the change).

However, this introduces a new deadlock:

- Thread 2 acquires the config_proxy lock and then removes an observer. It blocks waiting for the observer's CallGate to close.

- Thread 1 had dropped the config_proxy lock while traversing the observers to call each observer's handle_conf_change method. Those methods may attempt to reacquire the config_proxy lock. This creates the deadlock as it's waiting for Thread 2 to drop the lock while Thread 1 cannot release the CallGate.

The solution, I believe, is to properly refcount "uses" of the observers for the purposes of flushing these changes. Use std::shared_ptr to effect this.

Reproducing this is fairly simply with several parallel calls to `config set`. During the course of executing `config set`, the Objecter may receive config updates that will be flushed and potentially race with cleanup of observers during shutdown.

Fixes: https://tracker.ceph.com/issues/62832
Partial-revert: 7e8c683
Partial-revert: 4458a72

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by batrick** [comment]
> Reproducing this is fairly simply with several parallel calls to config set. During the course of executing config set, the Objecter may receive config updates that will be flushed and potentially race with cleanup of observers during shutdown.

I will add a reproducer to the QA suite but I wanted to get early feedback as well.

**Comment by batrick** [comment]
@vshankar @rzarzynski this should be good to go.

@athanatos I added some missing mutex locks. Do any/all of them need wrapped in crimson ifdefs? It seems only a few methods which acquire the lock actually required an ifdef.

**Comment by batrick** [comment]
I think I'm done shaking out the jenkins failures. This PR should be ready for review/QA.

**Comment by batrick** [comment]
https://tracker.ceph.com/projects/cephfs/wiki/Main#16-Oct-2023

Passes testing in `fs` suite.

ping @rzarzynski @athanatos @vshankar

**Comment by batrick** [comment]
https://tracker.ceph.com/projects/cephfs/wiki/Main#24-October-2023

File: src/common/config_proxy.h [hunk 1]
```diff
@@ -200,34 +160,39 @@ class ConfigProxy {
...
+    cond.notify_all();
   }
   void remove_observer(md_config_obs_t* obs) {
-    std::lock_guard l(lock);
-    call_gate_close(obs);
-    obs_call_gate.erase(obs);
-    obs_mgr.remove_observer(obs);
+    std::unique_lock l(lock);
+    auto wptr = obs_mgr.remove_observer(obs);
+    while (!wptr.expired()) {
```
Comment (vshankar) [inline src/common/config_proxy.h:181]: 👍

**Review by vshankar — APPROVED** [review]
LGTM. Fantastic work @batrick

**Review by Matan-B — APPROVED** [review]
Crimson change LGTM

**Pull request merged**